processor/src/main/java/org/mapstruct/ap/internal/conversion/JavaStreamToCollectionConversion.java

```diff
@@ -0,0 +1,95 @@
+/**
+ *  Copyright 2012-2016 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.internal.conversion;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.NavigableSet;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import org.mapstruct.ap.internal.model.common.ConversionContext;
+import org.mapstruct.ap.internal.model.common.Type;
+import org.mapstruct.ap.internal.model.common.TypeFactory;
+import org.mapstruct.ap.internal.util.Collections;
+import org.mapstruct.ap.internal.util.JavaStreamConstants;
+
+/**
+ * Conversion between {@link java.util.stream.Stream} and java utils: {@link Set}, {@link java.util.List},
+ * {@link java.util.Collection}.
+ *
+ * @author Filip Hrisafov
+ */
+public class JavaStreamToCollectionConversion extends SimpleConversion {
+
+    @Override
+    protected String getToExpression(ConversionContext conversionContext) {
+        return "Optional.ofNullable( <SOURCE> )" +
+            ".map( s -> s.collect( Collectors." + collector( conversionContext ) + " ) )" +
+            ".orElse( null )";
+    }
+
+    private String collector(ConversionContext conversionContext) {
+        Type targetType = conversionContext.getTargetType();
+        TypeFactory typeFactory = conversionContext.getTypeFactory();
+        String collector = "";
+        // we must do erasure as JDK compiler does not work without it
+        //TODO is this a bug in the Java compiler or not?
+        if ( typeFactory.getType( Iterable.class ).isAssignableTo( targetType.erasure() )
+            || typeFactory.getType( Collection.class ).isAssignableTo( targetType.erasure() )
+            || typeFactory.getType( List.class ).isAssignableTo( targetType.erasure() ) ) {
+            collector = "toList()";
+        }
+        else if ( typeFactory.getType( Set.class ).isAssignableTo( targetType.erasure() ) ) {
+            collector = "toSet()";
+        }
+        else if ( acceptsTreeSet( conversionContext ) ) {
+            collector = "toCollection( TreeSet::new )";
+        }
+
+        //TODO what if we register wrong. In the end it won't compile, but should we throw an exception?
```

Comment:
We will already raise such error, as no conversion could be found before. The only reason for something like this to happen is if we register a new type and forget to update the `JavaStreamToCollectionConversion`. In order to avoid such things, we could maybe do something in the register methods and make sure that the providers supports the conversion that we are trying to register.